script.adzapper/detect.py

```diff
@@ -0,0 +1,89 @@
+# -*- coding: utf-8 -*-
+import xbmc
+import xbmcgui
+import xbmcaddon
+
+# Import the common settings
+from resources.lib.settings import log
+from resources.lib.settings import Settings
+
+ADDON = xbmcaddon.Addon(id='script.adzapper')
+CWD = ADDON.getAddonInfo('path').decode("utf-8")
+
+
+#######################################
+# Window to detect a remote key press
+#######################################
+class DetectWindow(xbmcgui.WindowXMLDialog):
+    BACKGROUND_IMAGE = 101
+    TEXT_LABEL = 103
+    CANCEL_BUTTON = 104
+
+    def __init__(self, *args, **kwargs):
+        self.closed = False
+        self.buttonCode = None
+
+        xbmcgui.WindowXMLDialog.__init__(self)
+
+    @staticmethod
+    def createDetectWindow():
+        return DetectWindow("script-adzapper-detect.xml", CWD)
+
+    # Called when setting up the window
+    def onInit(self):
+        # Set the labels on the decrease and increase buttons
+        imageControl = self.getControl(DetectWindow.BACKGROUND_IMAGE)
+        imageControl.setImage(ADDON.getAddonInfo('fanart'))
+
+        xbmcgui.WindowXMLDialog.onInit(self)
+
+    # Handle the close action
+    def onAction(self, action):
+        ACTION_PREVIOUS_MENU = 10
+        ACTION_NAV_BACK = 92
+        if (action.getId() == ACTION_PREVIOUS_MENU) or (action.getId() == ACTION_NAV_BACK):
+            log("DetectWindow: Close Action received: %s" % str(action.getId()))
+            self.close()
+        elif action.getButtonCode() not in [None, "", 0]:
+            labelControl = self.getControl(DetectWindow.TEXT_LABEL)
+            labelControl.setText(ADDON.getLocalizedString(32103))
+            self.buttonCode = str(action.getButtonCode())
+            log("DetectWindow: Button Code is %s" % self.buttonCode)
+
+            # Need to make a call to the settings to enable the new value
+            Settings.setKeymapData(self.buttonCode)
+
+            # Now leave a little time before we continue, the service should pick
+            # up the settings change and create a new mapping file
+            monitor = xbmc.Monitor()
```

Comment:
this monitor instance should probably be moved out of the class or to a class member variable, you're instantiating a new one and removing with every action performed on the dialog

Reply:
You are right but this is the script for detecting a button and will just be launched from settings (xml) here:

`
<setting visible="eq(-2,3) + eq(-1,true)" label="32025" option="close" type="action" action="RunScript($CWD/detect.py)"/>
`

So just once the user is starting button detection (which sadly is not allowed).
Mus think about this and maybe remove the whole mapping code.